File: go/spore-core/execution_scaffold_test.go
package sporecore

import (
	"context"
	"testing"
)

// Composable Execution runtime scaffold tests (issue #123): BudgetContext.Charge
// / Remaining / ContinuesRemaining, StrategyOutcome variant discrimination,
// SpanStack / BudgetStack push/pop, recursive *ExecutionContext threading, and
// distinct sibling BudgetContexts. Runtime-only types — no fixtures.

func totalSteps(value uint32) BudgetPolicy {
	return BudgetPolicy{Kind: BudgetTotalSteps, Value: value}
}

func unlimited() BudgetPolicy {
	return BudgetPolicy{Kind: BudgetUnlimited}
}

func failBehavior() BudgetExhaustedBehavior {
	return BudgetExhaustedBehavior{Kind: BehaviorFail}
}

func escalateBehavior() BudgetExhaustedBehavior {
	return BudgetExhaustedBehavior{Kind: BehaviorEscalate}
}

func continueBehavior(maxContinues uint32) BudgetExhaustedBehavior {
	return BudgetExhaustedBehavior{
		Kind:         BehaviorContinue,
		MaxContinues: maxContinues,
		OnExhausted:  &BudgetExhaustedBehavior{Kind: BehaviorFail},
	}
}

// ---- BudgetContext.Charge --------------------------------------------------

func TestChargeWithinAllowanceIncrementsStepsTaken(t *testing.T) {
	cx := NewBudgetContext(totalSteps(5), failBehavior(), "p")
	if exhausted := cx.Charge(2); exhausted != nil {
		t.Fatalf("charge(2) within allowance signalled exhaustion: %+v", exhausted)
	}
	if cx.StepsTaken != 2 {
		t.Fatalf("StepsTaken = %d, want 2", cx.StepsTaken)
	}
	if exhausted := cx.Charge(3); exhausted != nil {
		t.Fatalf("charge(3) to exactly the cap signalled exhaustion: %+v", exhausted)
	}
	if cx.StepsTaken != 5 {
		t.Fatalf("StepsTaken = %d, want 5", cx.StepsTaken)
	}
}

func TestChargeOverflowSignalsExhaustionWithoutMutation(t *testing.T) {
	cx := NewBudgetContext(totalSteps(3), continueBehavior(2), "phaseX")
	if exhausted := cx.Charge(2); exhausted != nil {
		t.Fatalf("charge(2) signalled exhaustion: %+v", exhausted)
	}
	// StepsTaken == 2; charging 2 more would exceed the cap of 3.
	exhausted := cx.Charge(2)
	if exhausted == nil {
		t.Fatal("charge overflow did not signal exhaustion")
	}
	// No mutation on overflow.
	if cx.StepsTaken != 2 {
		t.Fatalf("StepsTaken mutated on overflow: %d, want 2", cx.StepsTaken)
	}
	// Signal captures current state.
	if exhausted.StepsTaken != 2 {
		t.Fatalf("signal StepsTaken = %d, want 2", exhausted.StepsTaken)
	}
	if exhausted.Phase != "phaseX" {
		t.Fatalf("signal Phase = %q, want phaseX", exhausted.Phase)
	}
	if exhausted.Policy.Kind != BudgetTotalSteps || exhausted.Policy.Value != 3 {
		t.Fatalf("signal Policy = %+v, want total_steps/3", exhausted.Policy)
	}
	if exhausted.Behavior.Kind != BehaviorContinue {
		t.Fatalf("signal Behavior.Kind = %q, want continue", exhausted.Behavior.Kind)
	}
	if exhausted.ContinuesUsed != 0 {
		t.Fatalf("signal ContinuesUsed = %d, want 0", exhausted.ContinuesUsed)
	}
}

func TestChargeUnlimitedNeverExhausts(t *testing.T) {
	cx := NewBudgetContext(unlimited(), failBehavior(), "u")
	for i := 0; i < 1000; i++ {
		if exhausted := cx.Charge(1_000_000); exhausted != nil {
			t.Fatalf("unlimited charge signalled exhaustion at iter %d", i)
		}
	}
}

// ---- BudgetContext.Remaining -----------------------------------------------

func TestRemainingCappedAndUnlimited(t *testing.T) {
	capped := NewBudgetContext(totalSteps(5), failBehavior(), "c")
	if rem, ok := capped.Remaining(); !ok || rem != 5 {
		t.Fatalf("fresh capped Remaining = (%d,%v), want (5,true)", rem, ok)
	}
	_ = capped.Charge(2)
	if rem, ok := capped.Remaining(); !ok || rem != 3 {
		t.Fatalf("after charge(2) Remaining = (%d,%v), want (3,true)", rem, ok)
	}
	// Saturates at 0, never negative: charge to the cap then probe.
	_ = capped.Charge(3)
	if rem, ok := capped.Remaining(); !ok || rem != 0 {
		t.Fatalf("at cap Remaining = (%d,%v), want (0,true)", rem, ok)
	}

	unl := NewBudgetContext(unlimited(), failBehavior(), "u")
	if rem, ok := unl.Remaining(); ok {
		t.Fatalf("unlimited Remaining = (%d,%v), want (_, false)", rem, ok)
	}
}

// ---- BudgetContext.ContinuesRemaining --------------------------------------

func TestContinuesRemainingForContinueEscalateFail(t *testing.T) {
	cont := NewBudgetContext(unlimited(), continueBehavior(3), "c")
	if got := cont.ContinuesRemaining(); got != 3 {
		t.Fatalf("fresh Continue ContinuesRemaining = %d, want 3", got)
	}
	cont.ContinuesUsed = 1
	if got := cont.ContinuesRemaining(); got != 2 {
		t.Fatalf("after 1 used ContinuesRemaining = %d, want 2", got)
	}
	// Saturates at 0, never negative.
	cont.ContinuesUsed = 5
	if got := cont.ContinuesRemaining(); got != 0 {
		t.Fatalf("over-used ContinuesRemaining = %d, want 0 (saturating)", got)
	}

	esc := NewBudgetContext(unlimited(), escalateBehavior(), "e")
	if got := esc.ContinuesRemaining(); got != 0 {
		t.Fatalf("Escalate ContinuesRemaining = %d, want 0", got)
	}
	fail := NewBudgetContext(unlimited(), failBehavior(), "f")
	if got := fail.ContinuesRemaining(); got != 0 {
		t.Fatalf("Fail ContinuesRemaining = %d, want 0", got)
	}
}

// ---- StrategyOutcome variant discrimination --------------------------------

func TestStrategyOutcomeVariantDiscrimination(t *testing.T) {
	partial := "half"
	exhausted := StrategyBudgetExhausted(BudgetExhausted{
		Policy:     totalSteps(2),
		Behavior:   failBehavior(),
		StepsTaken: 2,
		Phase:      "x",
	}, &partial)
	failed := StrategyFailed(&StrategyNotFoundError{Key: "k"})
	complete := StrategyComplete("done")

	// A BudgetExhausted is NOT a Failed — callers can distinguish.
	if exhausted.Kind != StrategyOutcomeBudgetExhausted {
		t.Fatalf("exhausted.Kind = %q, want budget_exhausted", exhausted.Kind)
	}
	if exhausted.Kind == StrategyOutcomeFailed {
		t.Fatal("BudgetExhausted collapsed into Failed")
	}
	if exhausted.Exhausted == nil || exhausted.Exhausted.PartialOutput == nil || *exhausted.Exhausted.PartialOutput != "half" {
		t.Fatalf("exhausted payload/partial output wrong: %+v", exhausted.Exhausted)
	}
	if exhausted.Exhausted.StepsTaken != 2 {
		t.Fatalf("exhausted StepsTaken = %d, want 2", exhausted.Exhausted.StepsTaken)
	}

	if failed.Kind != StrategyOutcomeFailed {
		t.Fatalf("failed.Kind = %q, want failed", failed.Kind)
	}
	if failed.Failed == nil {
		t.Fatal("Failed outcome carries no error")
	}
	if complete.Kind != StrategyOutcomeComplete || complete.Complete != "done" {
		t.Fatalf("complete outcome wrong: %+v", complete)
	}
}

// ---- SpanStack push/pop depth ----------------------------------------------

func TestSpanStackPushPopDepth(t *testing.T) {
	var spans SpanStack
	if spans.Depth() != 0 {
		t.Fatalf("fresh SpanStack depth = %d, want 0", spans.Depth())
	}
	spans.Push("a")
	spans.Push("b")
	if spans.Depth() != 2 {
		t.Fatalf("depth after 2 pushes = %d, want 2", spans.Depth())
	}
	id, ok := spans.Pop()
	if !ok || id != "b" {
		t.Fatalf("pop = (%q,%v), want (b,true)", id, ok)
	}
	if spans.Depth() != 1 {
		t.Fatalf("depth after pop = %d, want 1", spans.Depth())
	}
	_, _ = spans.Pop()
	if _, ok := spans.Pop(); ok {
		t.Fatal("pop on empty SpanStack returned ok=true")
	}
}

// ---- BudgetStack push/pop --------------------------------------------------

func TestBudgetStackPushPopCurrent(t *testing.T) {
	var budgets BudgetStack
	if budgets.Current() != nil {
		t.Fatal("empty BudgetStack Current() != nil")
	}
	if _, ok := budgets.Pop(); ok {
		t.Fatal("pop on empty BudgetStack returned ok=true")
	}
	budgets.Push(NewBudgetContext(totalSteps(2), failBehavior(), "outer"))
	budgets.Push(NewBudgetContext(totalSteps(4), failBehavior(), "inner"))
	if budgets.Depth() != 2 {
		t.Fatalf("depth = %d, want 2", budgets.Depth())
	}
	// Current() is mutable and points at the innermost scope.
	cur := budgets.Current()
	if cur.Phase != "inner" {
		t.Fatalf("Current().Phase = %q, want inner", cur.Phase)
	}
	_ = cur.Charge(1)
	if budgets.Current().StepsTaken != 1 {
		t.Fatal("Current() did not yield a mutable pointer into the stack")
	}
	popped, ok := budgets.Pop()
	if !ok || popped.Phase != "inner" {
		t.Fatalf("pop = (%q,%v), want (inner,true)", popped.Phase, ok)
	}
	if budgets.Current().Phase != "outer" {
		t.Fatalf("after pop Current().Phase = %q, want outer", budgets.Current().Phase)
	}
}

// ---- recursive stub strategy threads *ExecutionContext + BudgetStack --------

// recursiveStub recurses depth times, pushing a fresh per-node BudgetContext on
// descent and popping it on ascent, so the stack returns to baseline. Each node
// (incl. siblings at the same level) gets its OWN BudgetContext.
type recursiveStub struct {
	depth    int
	maxDepth *int
}

func (s recursiveStub) Run(ctx context.Context, cx *ExecutionContext) StrategyOutcome {
	cx.Budgets.Push(NewBudgetContext(totalSteps(uint32(s.depth+1)), failBehavior(), "node"))
	if d := cx.Budgets.Depth(); d > *s.maxDepth {
		*s.maxDepth = d
	}
	if s.depth > 0 {
		// Two children per node — siblings share the parent context but each
		// pushes/pops its own BudgetContext.
		_ = recursiveStub{depth: s.depth - 1, maxDepth: s.maxDepth}.Run(ctx, cx)
		_ = recursiveStub{depth: s.depth - 1, maxDepth: s.maxDepth}.Run(ctx, cx)
	}
	cx.Budgets.Pop()
	return StrategyComplete("")
}

func TestRecursiveStubThreadsContextAndBudgetStack(t *testing.T) {
	registry := NewExecutionRegistryBuilder().Build()
	cx := NewExecutionContext(&registry)
	maxDepth := 0
	outcome := recursiveStub{depth: 3, maxDepth: &maxDepth}.Run(context.Background(), cx)

	if outcome.Kind != StrategyOutcomeComplete {
		t.Fatalf("recursive stub outcome = %q, want complete", outcome.Kind)
	}
	// Stack returned to baseline depth (every push had a matching pop).
	if cx.Budgets.Depth() != 0 {
		t.Fatalf("BudgetStack depth after run = %d, want 0", cx.Budgets.Depth())
	}
	// Recursion actually descended (depth 3 → at least 4 frames live at once).
	if maxDepth < 4 {
		t.Fatalf("max live BudgetStack depth = %d, want >= 4", maxDepth)
	}
	// The same registry pointer was threaded throughout.
	if cx.Registry != &registry {
		t.Fatal("ExecutionContext.Registry pointer changed during recursion")
	}
}

// ---- siblings get distinct BudgetContexts ----------------------------------

func TestSiblingsGetDistinctBudgetContexts(t *testing.T) {
	var budgets BudgetStack
	budgets.Push(NewBudgetContext(totalSteps(5), failBehavior(), "sibling-a"))
	// Charge the first sibling, then pop it.
	_ = budgets.Current().Charge(3)
	if budgets.Current().StepsTaken != 3 {
		t.Fatalf("sibling-a StepsTaken = %d, want 3", budgets.Current().StepsTaken)
	}
	budgets.Pop()

	// A second sibling at the same level starts fresh — no shared counters.
	budgets.Push(NewBudgetContext(totalSteps(5), failBehavior(), "sibling-b"))
	if budgets.Current().StepsTaken != 0 {
		t.Fatalf("sibling-b StepsTaken = %d, want 0 (siblings must not share)", budgets.Current().StepsTaken)
	}
	if budgets.Current().Phase != "sibling-b" {
		t.Fatalf("sibling-b Phase = %q, want sibling-b", budgets.Current().Phase)
	}
}

File: typescript/packages/core/tests/execution-scaffold.test.ts
/**
 * Composable Execution runtime scaffold (spore-core issue #123).
 *
 * StrategyOutcome + ExecutionContext / BudgetContext / BudgetStack / SpanStack.
 * Scaffold-only slice: these tests exercise the type surface and the threading,
 * NOT enforcement semantics. They mirror the Rust `execution_scaffold_tests`
 * module in `rust/crates/spore-core/src/harness.rs`:
 *   - charge within allowance increments stepsTaken,
 *   - overflow returns the error with correct fields and no mutation,
 *   - unlimited never exhausts,
 *   - remaining() capped + unlimited,
 *   - continuesRemaining() for continue / escalate / fail,
 *   - StrategyOutcome budget_exhausted vs failed discrimination,
 *   - SpanStack push/pop depth,
 *   - a recursive stub threads the shared ExecutionContext + BudgetStack and
 *     returns to baseline depth,
 *   - siblings get distinct BudgetContexts.
 */

import { describe, expect, it } from "vitest";

import {
  BudgetContext,
  BudgetStack,
  ExecutionRegistry,
  InvalidConfiguration,
  SpanStack,
  newExecutionContext,
  type BudgetExhaustedBehavior,
  type BudgetPolicy,
  type ExecutionContext,
  type RunStrategy,
  type StrategyOutcome,
} from "../src/index.js";
import { SpanId } from "../src/observability/types.js";

function continueBehavior(maxContinues: number): BudgetExhaustedBehavior {
  return { kind: "continue", max_continues: maxContinues, on_exhausted: { kind: "fail" } };
}

const fail: BudgetExhaustedBehavior = { kind: "fail" };
const escalate: BudgetExhaustedBehavior = { kind: "escalate" };

// ── BudgetContext.charge ─────────────────────────────────────────────────────

describe("BudgetContext.charge", () => {
  it("within allowance increments stepsTaken", () => {
    const cx = new BudgetContext({ kind: "per_loop", value: 5 }, fail, "loop");
    expect(cx.charge(2).ok).toBe(true);
    expect(cx.stepsTaken).toBe(2);
    expect(cx.charge(3).ok).toBe(true); // exactly to the cap
    expect(cx.stepsTaken).toBe(5);
    expect(cx.remaining()).toBe(0);
  });

  it("overflow returns the error with correct fields and no mutation", () => {
    const policy: BudgetPolicy = { kind: "total_steps", value: 3 };
    const behavior = continueBehavior(2);
    const cx = new BudgetContext(policy, behavior, "phase-x");
    expect(cx.charge(3).ok).toBe(true);

    const res = cx.charge(1);
    expect(res.ok).toBe(false);
    if (res.ok) throw new Error("expected exhaustion");
    expect(res.error).toEqual({
      policy,
      behavior,
      stepsTaken: 3,
      continuesUsed: 0,
      phase: "phase-x",
    });
    // state unchanged on failure
    expect(cx.stepsTaken).toBe(3);
  });

  it("unlimited never exhausts", () => {
    const cx = new BudgetContext({ kind: "unlimited" }, fail, "root");
    expect(cx.charge(Number.MAX_SAFE_INTEGER).ok).toBe(true);
    expect(cx.charge(100).ok).toBe(true);
    expect(cx.remaining()).toBeUndefined();
  });
});

// ── BudgetContext.remaining ──────────────────────────────────────────────────

describe("BudgetContext.remaining", () => {
  it("is capped for a bounded policy and undefined for unlimited", () => {
    const capped = new BudgetContext({ kind: "per_attempt", value: 10 }, fail, "attempt");
    expect(capped.remaining()).toBe(10);
    capped.charge(4);
    expect(capped.remaining()).toBe(6);

    const unlimited = new BudgetContext({ kind: "unlimited" }, fail, "u");
    expect(unlimited.remaining()).toBeUndefined();
  });
});

// ── BudgetContext.continuesRemaining ─────────────────────────────────────────

describe("BudgetContext.continuesRemaining", () => {
  it("counts continues for `continue` and is 0 for escalate / fail", () => {
    const cont = new BudgetContext({ kind: "unlimited" }, continueBehavior(3), "c");
    expect(cont.continuesRemaining()).toBe(3);
    cont.continuesUsed = 2;
    expect(cont.continuesRemaining()).toBe(1);
    cont.continuesUsed = 5; // saturates
    expect(cont.continuesRemaining()).toBe(0);

    const esc = new BudgetContext({ kind: "unlimited" }, escalate, "e");
    expect(esc.continuesRemaining()).toBe(0);

    const f = new BudgetContext({ kind: "unlimited" }, fail, "f");
    expect(f.continuesRemaining()).toBe(0);
  });
});

// ── StrategyOutcome discrimination ───────────────────────────────────────────

describe("StrategyOutcome", () => {
  it("budget_exhausted is distinguishable from failed and complete", () => {
    const exhausted: StrategyOutcome = {
      kind: "budget_exhausted",
      policy: { kind: "per_loop", value: 1 },
      behavior: fail,
      stepsTaken: 1,
      continuesUsed: 0,
      phase: "p",
      partialOutput: "partial",
    };
    const failed: StrategyOutcome = {
      kind: "failed",
      error: new InvalidConfiguration("boom"),
    };
    const complete: StrategyOutcome = { kind: "complete", output: "done" };

    // A budget_exhausted is NOT a failed — callers can distinguish by `kind`.
    const kinds = [exhausted, failed, complete].map((o): string => o.kind);
    expect(kinds).toEqual(["budget_exhausted", "failed", "complete"]);
    expect(kinds).not.toContain("pending"); // the old #119 placeholder is gone
    if (exhausted.kind === "budget_exhausted") {
      expect(exhausted.partialOutput).toBe("partial");
    }
  });
});

// ── SpanStack ────────────────────────────────────────────────────────────────

describe("SpanStack", () => {
  it("push/pop tracks depth", () => {
    const spans = new SpanStack();
    expect(spans.depth()).toBe(0);
    spans.push(new SpanId("a"));
    spans.push(new SpanId("b"));
    expect(spans.depth()).toBe(2);
    expect(spans.pop()).toEqual(new SpanId("b"));
    expect(spans.depth()).toBe(1);
  });
});

// ── Recursive stub threading the shared ExecutionContext ─────────────────────

/**
 * A recursive stub: pushes its own scope, optionally recurses, then pops —
 * modeling how a combinator threads the shared context and gives each node
 * (incl. siblings) its OWN BudgetContext.
 */
class RecursiveStub implements RunStrategy {
  constructor(
    private readonly value: number,
    private readonly children: RecursiveStub[],
  ) {}

  async run(cx: ExecutionContext): Promise<StrategyOutcome> {
    const baseline = cx.budgets.depth();
    cx.budgets.push(new BudgetContext({ kind: "per_loop", value: this.value }, fail, "node"));
    // charge against this node's own scope
    cx.budgets.current()?.charge(1);
    for (const child of this.children) {
      await child.run(cx);
    }
    // pop our scope; depth returns to baseline
    cx.budgets.pop();
    expect(cx.budgets.depth()).toBe(baseline);
    return { kind: "complete", output: "" };
  }
}

describe("recursive stub threading", () => {
  it("threads one ExecutionContext + BudgetStack and returns to baseline depth", async () => {
    const cx = newExecutionContext(ExecutionRegistry.empty());
    expect(cx.budgets.depth()).toBe(0);

    const tree = new RecursiveStub(4, [new RecursiveStub(2, []), new RecursiveStub(3, [])]);

    const outcome = await tree.run(cx);
    expect(outcome.kind).toBe("complete");
    // stack fully unwound after the recursive run
    expect(cx.budgets.depth()).toBe(0);
    // the shared usage/session/spans are reachable through one context
    expect(cx.spans.depth()).toBe(0);
  });
});

// ── Siblings get distinct BudgetContexts ─────────────────────────────────────

describe("BudgetStack", () => {
  it("gives siblings distinct, non-shared BudgetContexts", () => {
    const budgets = new BudgetStack();
    budgets.push(new BudgetContext({ kind: "per_loop", value: 5 }, fail, "sib-a"));
    // mutate sibling A
    budgets.current()?.charge(2);
    const a = budgets.pop();
    expect(a?.stepsTaken).toBe(2);

    // sibling B starts fresh — no shared state with A
    budgets.push(new BudgetContext({ kind: "per_loop", value: 5 }, fail, "sib-b"));
    const b = budgets.current();
    expect(b?.stepsTaken).toBe(0);
    expect(a?.phase).not.toBe(b?.phase);
  });
});
